SteamKit2/SteamKit2/Util/Utils.cs

```diff
@@ -363,42 +365,95 @@ public static uint GetIPAddress( IPAddress ipAddr )
             return BitConverter.ToUInt32( addrBytes, 0 );
         }
 
-
-        public static uint EndianSwap( uint input )
+        public static IPAddress GetIPAddress( CMsgIPAddress ipAddr )
         {
-            return ( uint )IPAddress.NetworkToHostOrder( ( int )input );
+            if ( ipAddr.ShouldSerializev6() )
+            {
+                return new IPAddress( ipAddr.v6 );
+            }
+            else
+            {
+                return GetIPAddress( ipAddr.v4 );
+            }
         }
-        public static ulong EndianSwap( ulong input )
+
+        public static CMsgIPAddress GetMsgIPAddress( IPAddress ipAddr )
         {
-            return ( ulong )IPAddress.NetworkToHostOrder( ( long )input );
+            var msgIpAddress = new CMsgIPAddress();
+            byte[] addrBytes = ipAddr.GetAddressBytes();
+
+            if ( ipAddr.AddressFamily == AddressFamily.InterNetworkV6 )
+            {
+                msgIpAddress.v6 = addrBytes;
+            }
+            else
+            {
+                Array.Reverse( addrBytes );
+
+                msgIpAddress.v4 = BitConverter.ToUInt32( addrBytes, 0 );
+            }
+
+            return msgIpAddress;
         }
-        public static ushort EndianSwap( ushort input )
+
+        public static CMsgIPAddress ObfuscatePrivateIP( IPAddress ipAddr )
         {
-            return ( ushort )IPAddress.NetworkToHostOrder( ( short )input );
+            var localIp = GetMsgIPAddress( ipAddr );
+
+            if ( ipAddr.AddressFamily == AddressFamily.InterNetworkV6 )
+            {
+                // TODO: See ObfuscatePrivateIP in steamclient, unsure if applied correctly

```

Comment:
This looks right to me unless there's something funny about byte order that I've overlooked...

Reply:
So using the following sample code, to try mirror the function from steamclient:

```c++
int main()
{
	int8_t data[16];
	memset(data, 0, sizeof(data));

	auto i = 0;
	int32_t* ptr = (int32_t*)data;
	do
	{
		ptr[i] = ptr[i] ^ 0xBADF00D;

		i++;
	} while (i < 4);

	for (auto i = 0; i < sizeof(data); i++)
	{
		std::cout << data << std::endl;
	}
	return 0;
}
```

It looks like we actually need to XOR in reverse, i.e. `0x0D, 0xF0, 0xAD, 0xBA` as the mask. You're using the high byte to obfuscate the low byte and vice-versa.